js/webapp.js

```diff
@@ -927,26 +928,22 @@ WebApp.prototype = {
               utilLog.debug("After myProxy call");
             }
             else {
-              do404(req.url, res, this.options.productCode
+              utilLog.debug(`Referrer proxying miss. Resource not found, sending 404 because referrer (${referrer}) didn't match an existing proxy service`);
+              return do404(req.url, res, this.options.productCode
               + ": unknown resource requested");
             }
           }
-          else {
-            do404(req.url, res, this.options.productCode
-            + ": unknown resource requested");
+            else {
+              utilLog.debug(`Referrer proxying miss. Resource not found, sending 404 because referrer (${referrer}) didn't match a plugin pattern`);               
+            return do404(req.url, res, this.options.productCode
+            + ": unknown resource requested. Referrer="+referrer);
           }
+        } else {
```

Comment:
If you construct a proxy (external-type dataservice), the proxied content can request absolute paths that would exist outside of the namespace of the proxy, such as /foo rather than /ZLUX/plugins/myplugin/service/proxy/foo. When the request for /foo hits the zlux server, it tends to include a referrer which can make you reasonably confident that the request for a missing resource was actually intended to be made to the proxy.

Reply:
OK, I see why this is needed. I'd add a comment to the code explaining this. Do you have real examples of plugins that rely on this feature? 

Still, having a URL endpoint that behaves really differently depending of the referrer feels bizarre to me. I know it's common to allow or block a request depending on the referrer, but I've never encountered that is serves different content. I think, this can lead to problems with caching (plugin A calls /foo, we proxy to /whatever/com.example.foo, the browser remembers the response, then plugin B calls /foo expecting /something-else/com.qwerty.foo but will get the first response instead)

So here it's a situation where we actually would want to set all possible no-cache headers to disable all caching. Which could, in turn, affect the performance.